### PR TITLE
ENH: Infinite df approximation for Studentized Range PDF

### PR DESCRIPTION
**Pull request by DominicChm**
#### Reference issue

Adds on to gh-13732

#### What does this implement/fix?

This PR implements an approximate integrator for the Studentized Range Distribution's PDF with infinite degrees of freedom. This is necessary to avoid numerical overflows in the standard integrator when using very high `df` inputs.
It also patches an `np.clip` statement into the output of the Studentized Range CDF to prevent integrator errors from returning over 1.

#### Additional information
The integrand is implemented in Cython like the rest of the Studentized Range is. It was created from an already implemented approximation of the CDF by differentiating with respect to x.

CDF Approximation (already implemented):
image

PDF Approximation (This PR):
image

### CODE REVIEW

File: scipy/stats/tests/test_distributions.py [hunk 1]
```diff
@@ -5235,6 +5235,38 @@ def test_fitstart_valid(self):
...
+        res = stats.studentized_range.cdf(3, 10, np.inf)
+        res_finite = stats.studentized_range.cdf(3, 10, 99999)
+        assert_allclose(res, res_finite, atol=1e-4, rtol=1e-4)
+
+    def test_df_cutoff(self):
+        # Test that the CDF and PDF properly switch integrators at df=100,000.
+        # The infinite integrator should be different enough that it fails
+        # an allclose assertion. Also sanity check that using the same
+        # integrator does pass the allclose with a 1-df difference, which should
+        # be tiny.
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5254]:
```suggestion
        # integrator does pass the allclose with a 1-df difference, which
        # should be tiny.
```
`scipy/stats/tests/test_distributions.py:5253:80: E501 line too long (80 > 79 characters)`

**Comment by mdhaber** [comment]
Let's silence these while we're poking around.

```
FAILED ../../stats/tests/test_distributions.py::TestStudentizedRange::test_moment_against_mp[case_result1]
FAILED ../../stats/tests/test_distributions.py::TestStudentizedRange::test_moment_vectorization
```
See e.g. [here](https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=16072&view=logs&j=e3555eda-5afc-5d78-d2e9-cf9756f98375&t=220e105d-2d02-5073-9f3b-414490a3ddd5&l=1298).

Unless we've already tried it before, I think all we need is `with np.errstate(invalid='ignore')` around 5185 and 5224

File: scipy/stats/_continuous_distns.py [hunk 2]
```diff
@@ -9153,18 +9153,25 @@ def _single_moment(K, k, df):
...
     def _pdf(self, x, k, df):
-        cython_symbol = '_studentized_range_pdf'
 
         def _single_pdf(q, k, df):
-            log_const = _stats._studentized_range_pdf_logconst(k, df)
-            arg = [q, k, df, log_const]
-            usr_data = np.array(arg, float).ctypes.data_as(ctypes.c_void_p)
+            # The infinite form of the PDF is derived from the infinite
+            # CDF.
+            if df < 100000:
```
Comment (mdhaber) [inline scipy/stats/_continuous_distns.py:9160]: How did you choose the cutoff?

Comment (DominicChm) [reply scipy/stats/_continuous_distns.py:9160]: That was selected way back when we investigating the performance of the log version of the CDF. It seemed that 100,000 was a safe limit for the standard integrand. Problems usually occurred at around 300,000. Here's an example of the graphs:
k-19
 
Now that you mention, it might not be valid for the PDF... I'll do some testing to see if it's OK.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:9160]: Yeah, that would be good to see. Probably not so different, but worth a look.

Comment (DominicChm) [reply scipy/stats/_continuous_distns.py:9160]: Graphs are the PDF result vs df. 

The approximation switch over the entire range of df seems to work pretty well, and is barely even noticeable:
q2-k3

Zooming in on the switch region there's a clear transition, but it's not of huge consequence.
q2 1-k3

The main point of the approximation switch is to handle np.inf inputs. Degrees of freedom above 120 aren't even tabulated in most tables, so situations where this transition actually matters are probably few and far between.

Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:9160]: > Graphs are the PDF result vs df.

At fixed `x` and `k`?

> so situations where this transition actually matters are probably few and far between.

I totally agree, but we should still have some rationale for our choice, or at least we should characterize the choice we make.

A simple way to do this would be to repeat our experiment (6.2 [here](https://zenodo.org/record/5151976#.Yex2nP7MKUk)) except with `df` fixed at 100000. Would that be pretty easy?

File: scipy/stats/tests/test_distributions.py [hunk 3]
```diff
@@ -5221,7 +5225,12 @@ def test_cdf_against_r(self, r_case_result):
     def test_moment_vectorization(self):
         # Test moment broadcasting. Calls `_munp` directly because
         # `rv_continuous.moment` is broken at time of writing. See gh-12192
-        m = stats.studentized_range._munp([1, 2], [4, 5], [10, 11])
+
+        # Silence invalid value encountered warnings. Actual problems will be
+        # caught by the result comparison.
+        with np.errstate(invalid='ignore'):
+            m = stats.studentized_range._munp([1, 2], [4, 5], [10, 11])
```
Comment (mdhaber) [inline scipy/stats/tests/test_distributions.py:5232]: This is good, but I just got a new one in `test_compare_r`.
https://dev.azure.com/scipy-org/SciPy/_build/results?buildId=16112&view=logs&j=e3555eda-5afc-5d78-d2e9-cf9756f98375&t=220e105d-2d02-5073-9f3b-414490a3ddd5&l=1298

These are only happening in 32-bit builds, I think, so they don't worry me very much. We know that the integrand is very challenging numerically, and we did the best we could. I'm OK with warnings in the tests as long as the results are good.

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (DominicChm) [reply scipy/stats/_continuous_distns.py:9160]: Performed the experiment - After a bunch of testing seems like the integrator is bizarrely good at controlling error at high values of nu. Attached are plots of relative exact and approximate integrator error vs the reference, MP math implementation, as well as a plot of the "step" from the exact to approximate integrator.
result
result_step

Below is a MPMath convergence plot to verify that the reference implementation (probably) isn't problematic. It's representative of all convergence plots preformed at a variety of Q and K values.

File: scipy/stats/tests/test_distributions.py (hunk 3 above)
Comment (DominicChm) [reply scipy/stats/tests/test_distributions.py:5233]: Just added an ignore to test_against_R.

File: scipy/stats/_continuous_distns.py (hunk 2 above)
Comment (mdhaber) [reply scipy/stats/_continuous_distns.py:9786]: to help other reviewers:

> Attached are plots of relative exact and approximate integrator error vs the reference, MP math implementation,

Is this the MP math implementation of the infinite `df` approximation or the MP math implementation for finite `df` evaluated at `df = 100000`?

Also, "SciPy Exact" means the implementation of `studentized_range.pdf` in `main`, and "SciPy Approx" means the infinite `df` approximation added by this PR, right?

Comment (DominicChm) [reply scipy/stats/_continuous_distns.py:9786]: The MP math reference implementation is for the exact integral at `df=100000`, and your nomenclature is correct for exact/approx.

File: scipy/stats/_stats.pyx [hunk 4]
```diff
@@ -602,6 +602,15 @@ cdef double _studentized_range_pdf(int n, double[2] integration_var,
     return math.exp(log_terms) * math.pow(_Phi(s * q + z) - _Phi(z), k - 2)
 
 
+cdef double _studentized_range_pdf_asymptotic(double z, void *user_data) nogil:
+    # evaluates the integrand of equation (2) by Lund, Lund, page 205. [4]
+    # destined to be used in a LowLevelCallable
```
Comment (mdhaber) [inline scipy/stats/_stats.pyx:607]: Note quite, right? This is the PDF, so it's the derivative of the integrand w.r.t. Q, no? And we actually worked from the formula shown on Wikipedia, which is written a little differently, so it might help to explain how this is equivalent if equation 2 is the reference.

**Comment by mdhaber** [comment]
@steppi I know review time is scarce, but I wondered whether you'd be interested in this when you have a moment? Most of it is copy-paste replication of the existing studentized range distribution setup but with a new for an infinite degree of freedom approximation. The main question, I think, is when (at what value of the `df` parameter) we should transition to the infinite degree of freedom approximation. @DominicChm did a pretty careful numerical study to back up the choice, but I'd appreciate your thoughts.

**Comment by steppi** [comment]
> @steppi I know review time is scarce, but I wondered whether you'd be interested in this when you have a moment? Most of it is copy-paste replication of the existing studentized range distribution setup but with a new for an infinite degree of freedom approximation. The main question, I think, is when (at what value of the `df` parameter) we should transition to the infinite degree of freedom approximation. @DominicChm did a pretty careful numerical study to back up the choice, but I'd appreciate your thoughts.

I’ll take a look this weekend.

**Comment by steppi** [comment]
I haven’t forgotten about this. I’m working out a rigorous error analysis and will try to post a write up and complete my review tomorrow.

**Comment by mdhaber** [comment]
Wow! Ok!

**Comment by mdhaber** [comment]
@sreppi just thought I'd check in since it sounded like the review was imminent. I have forgotten to actually post a review that I've written before, so I thought that might have happened here : ) if you're held up otherwise, though, no problem.

**Comment by steppi** [comment]
Oh, sorry. I've had... a lot going on. I did start writing something up, but things ended up even trickier than I expected. I haven't had enough time to write up where I got stuck. I'll try to post something here tomorrow. It will help that we have $\LaTeX$ now.

**Comment by steppi** [comment]
Here's a dump of everything I've found out so far. 

The nature of the approximation is as follows. The pdf can be expressed as

$$f_{\mathrm{R}}(q;k,\nu) =
\frac{\sqrt{2\pi}k(k-1)\nu^{\nu/2}}{\Gamma(\nu/2)2^{\nu/2 - 1}}\int_{0}^{\infty}
s^{\nu-1}\phi(\sqrt{\nu}s)\left[\int_{-\infty}^{\infty}s\phi(z+qs)\phi(z)\left[\Phi(z +
qs) - \Phi(z)\right]^{k-2}\operatorname{d}z\right]\operatorname{d}s$$

which in turn can be expressed as

$$f_{\mathrm{R}}(q;k,\nu) = \int_{0}^{\infty}f_{\mathcal{N}}\left(s, \frac{\nu}{2},
1\right)g(s;, q, k)\operatorname{d}s$$

where 

$$g(s; q, k) = k(k-1)\int_{-\infty}^{\infty}s\phi(z+qs)\phi(z)\left[\Phi(z +
q) - \Phi(z)\right]^{k-2}\operatorname{d}z$$

and

$$f_{\mathcal{N}}\left(s,\frac{\nu}{2}, 1\right) =
\frac{\sqrt{2\pi}\nu^{\nu/2}}{\Gamma(\nu/2)2^{\nu/2 -
1}}s^{\nu-1}\phi(\sqrt{v}s) = \frac{\nu^{\nu/2}}{\Gamma(\nu/2)2^{\nu/2 -
1}}s^{\nu-1}\exp \left(-\frac{\nu}{2}s^2\right)$$

is the pdf of the [Nakagami distribution](https://en.wikipedia.org/wiki/Nakagami_distribution)  with parameters  $m=\frac{\nu}{2}$ and $\Omega = 1$.

The mean of this Nakagami distribution has asymptotic expansion

$$\mu = \frac{\Gamma\left(\nu / 2 + 1 / 2\right)}{\Gamma\left(\nu /
2\right)}\sqrt{\frac{2}{\nu}} \sim 1 - \frac{1}{4\nu} + O\left(\frac{1}{\nu^2}\right)$$

and the variance has asymptotic expansion

$$\sigma^2 = 1 - \mu^2 \sim \frac{1}{2\nu} + O\left(\frac{1}{\nu^2}\right)$$

Therefore as $\nu$ grows, this Nakagami distribution becomes like a Dirac delta centered at $s = 1$, and we should expect the approximation, 

$$f_{\mathrm{R}}(q;k,\infty) =
k(k-1)\int_{-\infty}^{\infty}\phi(z+q)\phi(z)\left[\Phi(z + q) -
\Phi(z)\right]^{k-2}\operatorname{d}z$$

Things are complicated because the Nakagami pdf $f_{\mathcal{N}$ is not compactly supported. Thus analyzing the error depends on the behavior of $g$ in the tails. 

A similar approximation holds for the CDF, except one replaces $g$ with

$$G(s; q, k) = k\int_{-\infty}^{\infty}\phi(z)\left[\Phi(z + qs) -
\Phi(z)\right]^{k-1}\operatorname{d}z$$

The nice thing about $G$ is that it is bounded as a function of $s$ (and $q$) for fixed $k$. This isn't always the case for $g$. In the worst case we have $g(s; 0, 2) = s$. Fortunately, when $k = 2$ like here, the relevant integral actually can be simplified a lot, and I think we could actually avoid this approximation entirely.

We can look separately at the error in an appropriately sized interval around $$s = 1$$, and then also the error in the tails. In the former case, the error should be small because we expect $g(s; q, k)$ to be close to $g(1; q, k)$. I'm leaving out a lot of details, but the error here should be $O\left(\frac{1}{\nu}\right)$. Where the constant is proportional to the second derivative of $g$ with respect to $s$. I haven't really looked into the behavior of this second derivative yet.

For the CDF, $G(s; q, k) <= k$, so the behavior in the tails isn't really a concern. I checked and couldn't find this approximation used anywhere in the literature for the PDF. I'm slightly concerned that it may not have been used because of the behavior in the tails. Perhaps this was just obvious to the old numerical gurus, and they didn't even think to mention it.

The left tail isn't really a problem since the support ends at $s = 0$ and $g$ goes to zero as $s$ goes to zero. The CDF of the Nakagami distribution is
$$F_{\mathcal{N}}\left(s, \nu/2, 1\right) = \frac{\gamma(\nu/2,\:\nu s^2/ 2)}{\Gamma(\nu / 2)}$$

as a function of $s$, the right tail seems to have leading order behavior that's slightly thicker than a normal distribution.

It isn't an intractable problem to proceed further, but I'm just strapped for time at the moment. I see a lot of variability in the box and whisker plot above showing the error of the approximation. This variation in error probably isn't random, but due to unfavorable values of $q$ and $k$.

**tldr**
I tried doing a rigorous error analysis, but it's gotten stalled out for lack of time. The error for the PDF approximation is likely to be worse than that for the CDF, and it should depend more on $q$ and $k$. Given that my error analysis has stalled, I would like to see some empirical results for how the error depends on $q$ and $k$. Also, the case $k = 2$ may be problematic, but fortunately the integral simplifies a lot in this case, so I think we can just sidestep this entirely if we like. Although I may be wrong that this case is problematic, so empirical error estimates would help.

**Comment by mdhaber** [comment]
Thanks @steppi! 
@DominicChm has included some empirical error information [above](https://github.com/scipy/scipy/pull/15444#discussion_r823109258). I'll let him clarify exactly what "SciPy Exact Error" (e.g. I think that's SciPy's existing PDF implementation compared against an `mpmath` implementation that does the same - finite DOF - integral?)  etc.... mean and how other shape parameters were sampled (e.g. range).

**Comment by steppi** [comment]
> Thanks @steppi! @DominicChm has included some empirical error information [above](https://github.com/scipy/scipy/pull/15444#discussion_r823109258). I'll let him clarify exactly what "SciPy Exact Error" (e.g. I think that's SciPy's existing PDF implementation compared against an `mpmath` implementation that does the same - finite DOF - integral?) etc.... mean and how other shape parameters were sampled (e.g. range).

Thanks. I saw the plots. I think you're right that `Scipy Exact` compares the exact integral to an `mpmath` reference implementation for the finite DOF integral and `Scipy Approx` compares the approximation to the `mpmath` finite DOF integral.

**Comment by mdhaber** [comment]
Ok, so if those points in the plots were generated by calculating the error for randomly generated inputs over a wide range, would it be convincing that the error is typically low? We could explain in the documentation what error users might expect.
Or were you hoping for other experiments - maybe investigating particular cases, like `k = 2`?

**Comment by steppi** [comment]
> Ok, so if those points in the plots were generated by calculating the error for randomly generated inputs over a wide range, would it be convincing that the error is typically low? We could explain in the documentation what error users might expect. Or were you hoping for other experiments - maybe investigating particular cases, like `k = 2`?

Is the code for these simulations available? If so, I could vet it and run out a few extra benchmarks this coming weekend. I just want to check if there are any critical spots where the approximation breaks down and the change ends up breaking working cases. If there are any cases that break, they could probably be handled separately.

**Comment by albert-lendbuzz** [comment]
@mdhaber Ithink I’m OK with tentatively merging this and just following up later with a more systematic investigation.

**Comment by steppi** [comment]
@mdhaber I think I'm OK with tentatively merging this and following up later with a more systematic analysis.

**Review by steppi — APPROVED** [review]
I'm OK with tentatively merging this. The implementation is clearly correct and the numerical examples demonstrate that it is working properly.  It's possible the approximation may become too inaccurate for the choice of cutoff at $\nu = 10^5$ for some values of $q$ and $k$, but on the whole this PR looks like an improvement. @DominicChm I'd like to do a follow up analysis to verify the approximation is still working for potentially problematic $q$ and $k$ that may be problematic. Can you share the code for your empirical investigation comparing to an `mpmath` reference implementation?
